src/lib/mina_ledger/sparse_ledger.ml

```diff
@@ -7,33 +7,77 @@ let of_ledger_root ledger =
   of_root ~depth:(Ledger.depth ledger) (Ledger.merkle_root ledger)
 
 let of_ledger_subset_exn (oledger : Ledger.t) keys =
-  let ledger = Ledger.copy oledger in
-  let locations = Ledger.location_of_account_batch ledger keys in
-  let non_empty_locations = List.filter_map ~f:snd locations in
-  let accounts = Ledger.get_batch ledger non_empty_locations in
-  let merkle_paths = Ledger.merkle_path_batch ledger non_empty_locations in
-  let sl, _, _ =
+  let locations = Ledger.location_of_account_batch oledger keys in
+  let num_new_accounts, non_empty_locations =
+    let num_new_accounts = ref 0 in
+    let non_empty_locations =
+      List.filter_map locations ~f:(fun (_key, location) ->
+          if Option.is_none location then incr num_new_accounts ;
+          location )
+    in
+    (!num_new_accounts, non_empty_locations)
+  in
+  let accounts = Ledger.get_batch oledger non_empty_locations in
+  let merkle_paths, empty_merkle_paths =
+    let next_location_exn loc = Option.value_exn (Ledger.Location.next loc) in
+    let empty_address depth =
+      Ledger.Addr.of_directions @@ List.init depth ~f:(fun _ -> Direction.Left)
+    in
+    let merkle_path_locations =
+      (let rec add_locations remaining last_filled merkle_path_locations =
+         if remaining > 0 then
+           let new_location =
+             match last_filled with
+             | None ->
+                 Ledger.Location.Account (empty_address (Ledger.depth oledger))
+             | Some last_filled ->
+                 next_location_exn last_filled
+           in
+           add_locations (remaining - 1) (Some new_location)
+             (new_location :: merkle_path_locations)
+         else merkle_path_locations
+       in
+       add_locations )
+        num_new_accounts
+        (Ledger.last_filled oledger)
+        non_empty_locations
+    in
+    let merkle_paths = Ledger.merkle_path_batch oledger merkle_path_locations in
+    (let rec pop_empties num_empties merkle_paths locations acc =
```

Comment:
Can't we just do a list.split here? We don't use the `location` of empty accounts later.
I think we could just have `List.split merkle_paths num_new_accounts` here.

Reply:
@ghost-not-in-the-shell do you mind implementing this change and committing to this PR?